<?xml version="1.0"?>

<robot name="delto_5f" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:include filename="$(find dg_description)/urdf/dg5f_right.xacro" />
    <xacro:include filename="$(find dg_description)/urdf/dg5f_left.xacro" />

    <link name="world" />
    <joint name="world_to_left_fixed" type="fixed">
        <parent link="world" />
        <child link="left_base_link" />
        <origin xyz="0 -0.15 0.1" rpy="0 0 0" />
    </joint>

    <joint name="world_to_right_fixed" type="fixed">
        <parent link="world" />
        <child link="right_base_link" />
        <origin xyz="0 0.15 0.1" rpy="0 0 0" />
    </joint>

    <ros2_control name="GazeboSimSystem" type="system">
        <hardware>
            <plugin>ign_ros2_control/IgnitionSystem</plugin>
        </hardware>

        <joint name="lj_dg_1_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_1_2">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_1_3">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_1_4">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 2 joints -->
        <joint name="lj_dg_2_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_2_2">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_2_3">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_2_4">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 3 joints -->
        <joint name="lj_dg_3_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_3_2">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_3_3">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_3_4">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 4 joints -->
        <joint name="lj_dg_4_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_4_2">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_4_3">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_4_4">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 5 joints -->
        <joint name="lj_dg_5_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_5_2">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_5_3">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="lj_dg_5_4">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_1_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_1_2">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_1_3">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_1_4">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 2 joints -->
        <joint name="rj_dg_2_1">
            <command_interface name="position">

            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_2_2">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_2_3">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_2_4">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 3 joints -->
        <joint name="rj_dg_3_1">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_3_2">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_3_3">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_3_4">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 4 joints -->
        <joint name="rj_dg_4_1">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_4_2">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_4_3">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_4_4">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <!-- Finger 5 joints -->
        <joint name="rj_dg_5_1">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_5_2">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_5_3">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>

        <joint name="rj_dg_5_4">
            <command_interface name="position">
                
                
            </command_interface>
            <state_interface name="position">
                <param name="initial_position">0.0</param>
            </state_interface>
            <state_interface name="velocity" />
            <state_interface name="effort" />
        </joint>
    </ros2_control>

    <gazebo>

        <gravity>false</gravity>
        <physics type="dart">  
            <max_step_size>0.0</max_step_size>  
            <real_time_factor>1.0</real_time_factor>
            <real_time_update_rate>250</real_time_update_rate>  
            <max_contacts>5</max_contacts>  

            <solver>
                <type>quick</type>
                <iters>10</iters>  
                <precon_iters>0</precon_iters>
                <sor>1.0</sor>
            </solver>
        </physics>

        <self_collide>0</self_collide>

        <visual>
            <shadows>false</shadows>
            <grid>false</grid>
        </visual>

        <collision_detector>ode</collision_detector>
        <constraints>
            <contact_max_correcting_vel>0.0</contact_max_correcting_vel>
            <cfm>0.0</cfm>
            <erp>0.0</erp>
        </constraints>

        <plugin name="ign_ros2_control::IgnitionROS2ControlPlugin"
            filename="libign_ros2_control-system.so">
            <robot_param_node>robot_state_publisher</robot_param_node>
            <robot_param>robot_description</robot_param>
            <parameters>$(find dg5f_gz)/config/dg5f_both_gz_controller.yaml</parameters>
            <ros>
                <namespace>/</namespace>
            </ros>
        </plugin>

    </gazebo>

</robot>
